File: 강의 자료/02-알고리즘/15~17-동적 프로그래밍/문제별 코드/1309-동물원/solution.java
import java.io.*;
import java.util.*;
import java.lang.Math;

public class Main {
    static FastReader scan = new FastReader();
    static StringBuilder sb = new StringBuilder();

    static int N;
    static final int MOD = 9901;
    static int[][] Dy;
    
    static void input(){
        N = scan.nextInt();
        Dy = new int[N + 1][3];
    }

    static void pro() {
        // 초기값 구하기
        Dy[1][0] = 1;
        Dy[1][1] = 1;
        Dy[1][2] = 1;
    
        // 점화식을 토대로 Dy 배열 채우기
        for (int i = 2; i <= N; i++) {
            for (int prev = 0; prev <= 2; prev++) {
                Dy[i][0] += Dy[i - 1][prev];
                Dy[i][0] %= MOD;
                for (int j = 0; j <= 1; j++) {
                    if ((prev & (1 << j)) != 0) continue;
                    Dy[i][1 << j] += Dy[i - 1][prev];
                    Dy[i][1 << j] %= MOD;
                }
            }
        }
        System.out.println((Dy[N][0] + Dy[N][1] + Dy[N][2]) % MOD);
    }

    public static void main(String[] args) {
        input();
        pro();
    }


    static class FastReader {
        BufferedReader br;
        StringTokenizer st;

        public FastReader() {
            br = new BufferedReader(new InputStreamReader(System.in));
        }

        public FastReader(String s) throws FileNotFoundException {
            br = new BufferedReader(new FileReader(new File(s)));
        }

        String next() {
            while (st == null || !st.hasMoreElements()) {
                try {
                    st = new StringTokenizer(br.readLine());
                } catch (IOException e) {
                    e.printStackTrace();
                }
            }
            return st.nextToken();
        }

        int nextInt() {
            return Integer.parseInt(next());
        }

        long nextLong() {
            return Long.parseLong(next());
        }

        double nextDouble() {
            return Double.parseDouble(next());
        }

        String nextLine() {
            String str = "";
            try {
                str = br.readLine();
            } catch (IOException e) {
                e.printStackTrace();
            }
            return str;
        }
    }
}

File: 강의 자료/02-알고리즘/15~17-동적 프로그래밍/문제별 코드/1309-동물원/solution.py
import sys
si = sys.stdin.readline
MOD = 9901

n = int(si())

dy = [[0, 0, 0] for _ in range(n)]

# 초기값 채우기
dy[0][0], dy[0][1], dy[0][2] = 1, 1, 1

# 점화식을 토대로 dy 배열 채우기
for i in range(1, n):
    for prev in range(0, 3):
        dy[i][0] += dy[i - 1][prev]
        dy[i][0] %= MOD
        for j in range(2):
            if prev & (1 << j): continue
            dy[i][1 << j] += dy[i - 1][prev]
            dy[i][1 << j] %= MOD

print(sum(dy[n - 1]) % MOD)
